File: skills/sci-plot/assets/case-packs/rf-0159/plot.py
from pathlib import Path
from collections import defaultdict, Counter
import json
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.patches import Polygon
from matplotlib.collections import PatchCollection
from matplotlib.colors import Normalize, LinearSegmentedColormap
from openpyxl import load_workbook

ROOT = Path(__file__).resolve().parent
mpl.rcParams.update({"font.family": ["Arial Unicode MS", "DejaVu Sans"], "font.size": 8.5, "axes.linewidth": 0.7})
with open(ROOT / "china_city.geojson", encoding="utf-8") as f:
    city_geo = json.load(f)
with open(ROOT / "china_province.geojson", encoding="utf-8") as f:
    province_geo = json.load(f)

def read_xlsx(path):
    ws = load_workbook(path, read_only=True, data_only=True).active
    rows = list(ws.iter_rows(values_only=True))
    h = [str(x) for x in rows[0]]
    return [dict(zip(h, r)) for r in rows[1:] if any(v is not None for v in r)]

def rings(geom):
    t, c = geom["type"], geom["coordinates"]
    if t == "Polygon":
        return [c[0]]
    if t == "MultiPolygon":
        return [p[0] for p in c]
    return []

def lines(geom):
    t, c = geom["type"], geom["coordinates"]
    if t == "LineString":
        return [c]
    if t == "MultiLineString":
        return c
    return []

def centroid(feature):
    rr = rings(feature["geometry"])
    if not rr:
        ll = lines(feature["geometry"])
        arr = np.array([p for line in ll for p in line], float)
        return tuple(arr.mean(axis=0)) if len(arr) else (np.nan, np.nan)
    ring = max(rr, key=len)
    arr = np.asarray(ring, float)
    x, y = arr[:, 0], arr[:, 1]
    cross = x * np.roll(y, -1) - np.roll(x, -1) * y
    area = cross.sum() / 2
    if abs(area) < 1e-12:
        return float(x.mean()), float(y.mean())
    cx = ((x + np.roll(x, -1)) * cross).sum() / (6 * area)
    cy = ((y + np.roll(y, -1)) * cross).sum() / (6 * area)
    return float(cx), float(cy)

def base(ax, city_fill=None, norm=None, cmap=None):
    patches, values = [], []
    for f in city_geo["features"]:
        name = str(f["properties"].get("name"))
        for ring in rings(f["geometry"]):
            patches.append(Polygon(np.asarray(ring), closed=True))
            values.append(np.nan if city_fill is None else city_fill.get(name, np.nan))
    if city_fill is None:
        coll = PatchCollection(patches, facecolor="#F0EEE7", edgecolor="#FFFFFF", linewidth=0.12, zorder=1)
    else:
        colors = [("#EFEDE6" if not np.isfinite(v) else cmap(norm(v))) for v in values]
        coll = PatchCollection(patches, facecolor=colors, edgecolor="#FFFFFF", linewidth=0.10, zorder=1)
    ax.add_collection(coll)
    for f in province_geo["features"]:
        for ring in rings(f["geometry"]):
            a = np.asarray(ring)
            ax.plot(a[:, 0], a[:, 1], color="#4D5354", lw=0.36, zorder=2)
        for line in lines(f["geometry"]):
            a = np.asarray(line)
            ax.plot(a[:, 0], a[:, 1], color="#4D5354", lw=0.36, zorder=2)
    ax.set_xlim(73, 136)
    ax.set_ylim(17, 54.5)
    ax.set_aspect("equal", adjustable="box")
    ax.set_xticks([])
    ax.set_yticks([])
    ax.spines[:].set_visible(False)

centroids = {str(f["properties"].get("name")): centroid(f) for f in city_geo["features"]}
d1 = read_xlsx(ROOT / "data1.xlsx")
d2 = read_xlsx(ROOT / "data2.xlsx")
by_city = defaultdict(lambda: {"total": 0.0, "genus": Counter()})
for r in d1:
    city = str(r["City"])
    n = float(r["Number of tick samples"] or 0)
    genus = str(r["Tick genus"])
    by_city[city]["total"] += n
    by_city[city]["genus"][genus] += n
genera = sorted({g for z in by_city.values() for g in z["genus"]})
palette = ["#B84C4C", "#D58B3E", "#708C48", "#318884", "#4D75A1", "#7B6599", "#AD617E", "#8B7355"]
gcol = {g: palette[i % len(palette)] for i, g in enumerate(genera)}
value_map = {str(r["City"]): float(r["Value"] or 0) for r in d2}
vals = np.array(list(value_map.values()), float)
norm = Normalize(vmin=0, vmax=np.percentile(vals, 98))
cmap = LinearSegmentedColormap.from_list("nf", ["#F4EADB", "#E9B66C", "#D8674A", "#7B2D45"])
fig, axes = plt.subplots(1, 2, figsize=(14.6, 7.3), facecolor="#FBFAF7")
base(axes[0])
available = [(city, z) for city, z in by_city.items() if city in centroids]
sizes = np.array([z["total"] for _, z in available])
for city, z in available:
    x, y = centroids[city]
    dominant = z["genus"].most_common(1)[0][0]
    s = 10 + 76 * np.sqrt(z["total"] / max(sizes.max(), 1))
    axes[0].scatter(x, y, s=s, c=gcol[dominant], edgecolor="white", linewidth=0.45, alpha=0.86, zorder=4)
axes[0].set_title("A  Tick sampling landscape", loc="left", fontsize=12, fontweight="bold")
axes[0].text(0.01, 0.01, "Bubble area: total tick samples\nColor: dominant tick genus", transform=axes[0].transAxes, fontsize=7, color="#606668", va="bottom")
handles = [plt.Line2D([0], [0], marker="o", linestyle="", markerfacecolor=gcol[g], markeredgecolor="white", label=g, markersize=5.5) for g in genera]
axes[0].legend(handles=handles, title="Dominant genus", loc="lower left", bbox_to_anchor=(0.0, 0.08), ncol=2, frameon=False, fontsize=6.3, title_fontsize=7)
base(axes[1], value_map, norm, cmap)
axes[1].set_title("B  City-level value distribution", loc="left", fontsize=12, fontweight="bold")
sm = mpl.cm.ScalarMappable(norm=norm, cmap=cmap)
cb = fig.colorbar(sm, ax=axes[1], fraction=0.028, pad=0.01, shrink=0.58)
cb.set_label("Value", fontsize=8)
cb.ax.tick_params(labelsize=7)
axes[1].text(0.01, 0.01, "Color scale capped at the 98th percentile\nfor legibility; source values unchanged", transform=axes[1].transAxes, fontsize=7, color="#606668", va="bottom")
fig.suptitle("China spatial evidence atlas", x=0.055, y=0.97, ha="left", fontsize=19, fontweight="bold", color="#202528")
fig.text(0.055, 0.925, "City geometries and province outlines are read directly from the supplied GeoJSON files", fontsize=9, color="#656B6E")
fig.text(0.055, 0.02, "Tick data: 196 of 198 named cities matched exactly; Wanzhou and Fuling districts were not positioned. Administrative boundaries are shown as supplied and imply no endorsement.", fontsize=7.2, color="#656B6E")
fig.tight_layout(rect=(0.025, 0.05, 0.99, 0.9), w_pad=1.5)
fig.savefig(ROOT / "plot_python.png", dpi=360, facecolor=fig.get_facecolor())
plt.close(fig)

File: skills/sci-plot/assets/case-packs/rf-0159/plot.R
args <- commandArgs(trailingOnly = FALSE)
file_arg <- sub("^--file=", "", args[grep("^--file=", args)])
root <- if (length(file_arg)) dirname(normalizePath(file_arg)) else getwd()
suppressPackageStartupMessages({library(sf); library(readxl); library(ggplot2); library(patchwork); library(ragg)})
city <- st_read(file.path(root, "china_city.geojson"), quiet = TRUE)
province <- st_read(file.path(root, "china_province.geojson"), quiet = TRUE)
city_poly <- city[grepl("POLYGON", as.character(st_geometry_type(city))), ]
city_line <- city[grepl("LINESTRING", as.character(st_geometry_type(city))), ]
province_poly <- province[grepl("POLYGON", as.character(st_geometry_type(province))), ]
province_line <- province[grepl("LINESTRING", as.character(st_geometry_type(province))), ]
d1 <- read_excel(file.path(root, "data1.xlsx"))
d2 <- read_excel(file.path(root, "data2.xlsx"))
agg <- aggregate(as.numeric(d1$`Number of tick samples`), list(City = d1$City, Genus = d1$`Tick genus`), sum, na.rm = TRUE)
names(agg)[3] <- "samples"
total <- aggregate(samples ~ City, agg, sum)
dom <- do.call(rbind, lapply(split(agg, agg$City), function(z) z[which.max(z$samples), c("City", "Genus")]))
ticks <- merge(total, dom, by = "City")
points <- suppressWarnings(st_point_on_surface(city))
points <- merge(points, ticks, by.x = "name", by.y = "City")
point_xy <- st_coordinates(points)
point_data <- cbind(st_drop_geometry(points), X = point_xy[, 1], Y = point_xy[, 2])
genera <- sort(unique(agg$Genus))
pal <- c("#B84C4C", "#D58B3E", "#708C48", "#318884", "#4D75A1", "#7B6599", "#AD617E", "#8B7355")
gcol <- setNames(rep(pal, length.out = length(genera)), genera)
base_theme <- theme_void(base_size = 9) + theme(plot.title = element_text(face = "bold", size = 12), plot.subtitle = element_text(size = 7, colour = "#606668"), plot.background = element_rect(fill = "#FBFAF7", colour = NA), panel.background = element_rect(fill = "#FBFAF7", colour = NA), legend.background = element_blank(), legend.key = element_blank())
p1 <- ggplot() + geom_sf(data = city_poly, fill = "#F0EEE7", colour = "white", linewidth = .06, show.legend = FALSE) + geom_sf(data = city_line, colour = "#D5D2CB", linewidth = .08, show.legend = FALSE) + geom_sf(data = province_poly, fill = NA, colour = "#4D5354", linewidth = .22, show.legend = FALSE) + geom_sf(data = province_line, colour = "#4D5354", linewidth = .22, show.legend = FALSE) + geom_point(data = point_data, aes(X, Y, size = samples, colour = Genus), alpha = .86) + scale_colour_manual(values = gcol, name = "Dominant genus") + scale_size_area(max_size = 7, guide = "none") + coord_sf(xlim = c(73, 136), ylim = c(17, 54.5), expand = FALSE) + labs(title = "A  Tick sampling landscape", subtitle = "Bubble area: total tick samples · color: dominant tick genus") + base_theme + theme(legend.position = c(.12, .2), legend.text = element_text(size = 6), legend.title = element_text(size = 7))
city2 <- merge(city, d2, by.x = "name", by.y = "City", all.x = TRUE)
city2_poly <- city2[grepl("POLYGON", as.character(st_geometry_type(city2))), ]
city2_line <- city2[grepl("LINESTRING", as.character(st_geometry_type(city2))), ]
cap <- as.numeric(quantile(d2$Value, .98, na.rm = TRUE))
legend_cols <- colorRampPalette(c("#F4EADB", "#E9B66C", "#D8674A", "#7B2D45"))(5)
p2 <- ggplot() + geom_sf(data = city2_poly, aes(fill = pmin(Value, cap)), colour = "white", linewidth = .05, show.legend = FALSE) + geom_sf(data = city2_line, colour = "#D5D2CB", linewidth = .08, show.legend = FALSE) + geom_sf(data = province_poly, fill = NA, colour = "#4D5354", linewidth = .22, show.legend = FALSE) + geom_sf(data = province_line, colour = "#4D5354", linewidth = .22, show.legend = FALSE) + annotate("rect", xmin = seq(75.8, 80.6, length.out = 5), xmax = seq(77, 81.8, length.out = 5), ymin = 18.6, ymax = 19.4, fill = legend_cols, colour = NA) + annotate("text", x = 76.4, y = 18.15, label = "0", hjust = .5, size = 2.1, colour = "#606668") + annotate("text", x = 81.2, y = 18.15, label = round(cap), hjust = .5, size = 2.1, colour = "#606668") + annotate("text", x = 78.8, y = 17.45, label = "Value", hjust = .5, size = 2.2, colour = "#606668") + scale_fill_gradientn(colours = c("#F4EADB", "#E9B66C", "#D8674A", "#7B2D45"), limits = c(0, cap * 1.000001), guide = "none") + coord_sf(xlim = c(73, 136), ylim = c(17, 54.5), expand = FALSE) + labs(title = "B  City-level value distribution", subtitle = "Color capped at the 98th percentile; source values unchanged") + base_theme
p <- p1 | p2
p <- p + plot_annotation(title = "China spatial evidence atlas", subtitle = "City geometries and province outlines are read directly from the supplied GeoJSON files", caption = "Tick data: 196 of 198 named cities matched exactly; Wanzhou and Fuling districts were not positioned. Administrative boundaries are shown as supplied and imply no endorsement.", theme = theme(plot.title = element_text(size = 19, face = "bold"), plot.subtitle = element_text(size = 9, colour = "#656B6E"), plot.caption = element_text(size = 7, colour = "#656B6E"), plot.background = element_rect(fill = "#FBFAF7", colour = NA)))
agg_png(file.path(root, "plot_r.png"), width = 14.6, height = 7.3, units = "in", res = 360, background = "#FBFAF7")
print(p)
dev.off()
